### PR TITLE
Prevent smoketest from clobbering the current virtualenv

### PR DESCRIPTION
**Pull request by hannes-ucsc**
smoketest installs the CLI into the currently active virtualenv or even the global site-dist if no virtualenv is active. The CLI's dependencies might differ from those of data-store creating potential breakage in data-store after smoketest.py runs. This PR changes smoketest to install the CLI into a temporary virtualenv and remove it afterwards.

### Test plan

Run smoketest.py manually.

### CODE REVIEW

File: tests/smoketest.py [hunk 1]
```diff
@@ -43,45 +42,58 @@ def run(command, runner=check_call, **kwargs):
     run("git clone --depth 1 --recurse-submodules https://github.com/HumanCellAtlas/data-store-cli")
 
 run("http --check-status https://${API_HOST}/v1/swagger.json > data-store-cli/swagger.json")
-run("pip install -r data-store-cli/requirements.txt")
-run("python -c 'import sys, hca.dss.regenerate_api as r; r.generate_python_bindings(sys.argv[1])' swagger.json",
-    cwd="data-store-cli")
-run("find data-store-cli/hca -name '*.pyc' -delete")
-run("pip install --upgrade --no-deps .", cwd="data-store-cli")
+venv = tempfile.mkdtemp()
```
Comment (ttung) [inline tests/smoketest.py]: use `with tempfile.TemporaryDirectory() as venv:`

it provides cleanup for free.

File: tests/smoketest.py [hunk 2]
```diff
@@ -43,45 +42,58 @@ def run(command, runner=check_call, **kwargs):
...
-        run("http -v --check-status https://${API_HOST}/v1/bundles/$(jq -r .bundle_uuid upload.json)?replica=gcp")
-        break
-    except SystemExit:
-        time.sleep(1)
-else:
-    parser.exit(RED("Failed to replicate bundle from AWS to GCP"))
-run("hca dss download --replica gcp $(jq -r .bundle_uuid upload.json)")
+    run(f"cat {bundle_dir}/sample.json | jq .uuid=env.sample_id | sponge {bundle_dir}/sample.json",
+        env=dict(os.environ, sample_id=sample_id))
+    run(f"{venv_bin}hca dss upload "
```
Comment (ttung) [inline tests/smoketest.py]: Does executing {venv_bin}/hca use that virtualenv even if one doesn't activate that venv?

Comment (hannes-ucsc) [reply tests/smoketest.py]: Yes. Running a binary from the venv's bin activates that venv on the fly for the duration of the process.

File: tests/smoketest.py (hunk 1 above)
Comment (hannes-ucsc) [reply tests/smoketest.py]: Will do.

**Comment by hannes-ucsc** [comment]
I piggy-backed edc3934 onto this PR to also prevent smoketest from dirtying the working copy. I can pull it out into another PR if requested.

File: .gitignore [hunk 3]
```diff
@@ -44,3 +44,5 @@ npm-debug.log
 
 # Local environment configuration
 /environment.local
+
+/smoketest-*.tmp
```
Comment (ttung) [inline .gitignore]: Why is this necessary since you have the `with tempfile.TemporaryDirectory...`

Comment (hannes-ucsc) [reply .gitignore]: I added `dir=` to that, placing the temp dir into the build directory. This makes the tempdir easier to find for post-mortems but I forgot to actually make clean-up optional by going back to  

```
cleanup=True
…
try:
    …
finally:
    if cleanup:
        rmdir(…)
```

Let me push that.

**Comment by hannes-ucsc** [comment]
Running smoketest now leaves no temp files lying around in the source tree, but can be optionally directed to skip cleaning up its temporary working directory (via the --no-clean options). This can be useful for post-mortems.

File: tests/smoketest.py [hunk 4]
```diff
@@ -39,49 +40,70 @@ def run(command, runner=check_call, **kwargs):
...
-        res = json.load(fh2)
-        print(json.dumps(res, indent=4))
-        assert len(res["results"]) == 1
-
-    res = run(f"hca dss put-subscriptions --callback-url https://example.com/ --es-query '{{}}' --replica {replica}",
-              runner=check_output)
-    sub_id = json.loads(res.decode())["uuid"]
-    run(f"hca dss get-subscriptions --replica {replica}")
-    run(f"hca dss delete-subscriptions --replica {replica} {sub_id}")
+workdir = tempfile.mkdtemp(dir=os.getcwd(), prefix="smoketest-", suffix='.tmp')
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply tests/smoketest.py]: I disagree. Considering the way the build currently pollutes the source tree with hundreds of files, smoketest would be the least of our problems if the source tree were on an NFS. IOW, what you propose as a fix for the NFS scenario doesn't go far enough anyway. So if I had to weigh that against the convenience of quickly finding smoketest's working directory after running with `--no-clean`, I'd chose the latter.

File: tests/smoketest.py [hunk 5]
```diff
@@ -39,49 +40,70 @@ def run(command, runner=check_call, **kwargs):
...
+                  "--callback-url https://example.com/ "
+                  "--es-query '{}' "
+                  f"--replica {replica}",
+                  runner=check_output)
+        sub_id = json.loads(res.decode())["uuid"]
+        run(f"{venv_bin}hca dss get-subscriptions --replica {replica}")
+        run(f"{venv_bin}hca dss delete-subscriptions --replica {replica} {sub_id}")
+finally:
+    if args.clean:
+        shutil.rmtree(workdir)
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply tests/smoketest.py]: Of course. I keep forgetting that we're not on Python 2.

File: tests/smoketest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ttung) [reply tests/smoketest.py]: We should build the deployment stuff out of directory as well, but I don't want to spend all day debating this.

Go ahead and make the `tempfile.TemporaryDirectory(..)` change and land this.

**Comment by hannes-ucsc** [comment]
@ttung's instructions were a little ambiguous but I resolved them in my favor, leaving `dir=` in place but changing `mkdtemp` to `TemporaryDirectory` and `shutil.rmtree(workdir)` to `workdir.cleanup`.
